SEP-1862: Tool Resolution

SamMorrowDrums opened the pull request:
## Summary

This SEP proposes a **tool resolution** mechanism that enables servers to provide argument-specific metadata before tool invocation. The design takes inspiration from LSP's resolve pattern and HTTP preflight OPTION requests while adapting it for MCP's unique requirements.

## Key Changes

- **New `tools/resolve` method**: Clients can request refined tool metadata based on intended arguments
- **New `resolve` field**: Tools can indicate they support resolution
- **New server capability**: `tools.resolve: true`

## The Problem

Tool definitions from `tools/list` are static. A `manage_files` tool must declare `destructiveHint: true` even though `read` operations are completely safe. This causes:
- Over-prompting with unnecessary confirmation dialogs
- Reduced trust from LLMs that avoid "destructive" tools
- Poor UX for versatile, multi-action tools

## The Solution

Before invoking a tool, clients can call `tools/resolve` with the tool name and intended arguments. The server returns a complete tool definition with annotations refined for that specific operation:

```json
// Request
{ "method": "tools/resolve", "params": { "name": "manage_files", "arguments": { "action": "read" }}}

// Response - safe operation!
{ "tool": { "name": "manage_files", "annotations": { "readOnlyHint": true, "destructiveHint": false }}}
```

## Why This Approach?

This design follows @findleyr's [suggestion](https://github.com/modelcontextprotocol/modelcontextprotocol/pull/1862#issuecomment-3573029197) to adopt an LSP-inspired resolve pattern:

> "we probably want to avoid a future where there are a bunch of separate preflight checks for a single tool call--it would be better to consolidate them into a single exchange"

By returning the full `Tool` object, we:
1. **Prevent method proliferation** (`tools/annotations`, `tools/scopes`, `tools/costs`)
2. **Enable future extensions** without protocol changes
3. **Provide a single preflight exchange** for all metadata needs

## Future Extensibility

The pattern supports future metadata beyond annotations:
- **Scope requirements**: OAuth scopes needed for specific operations
- **Cost estimates**: Token/credit costs that vary by arguments  
- **Rate limits**: Different operations may have different limits

These are explicitly out of scope for this SEP but the mechanism supports them.

## Backward Compatibility

Fully backward compatible:
- Existing servers work unchanged
- Existing clients can ignore `resolve: true`
- Static annotations remain the fallback

---

**Checklist:**
- [x] SEP follows the template format
- [x] Abstract and motivation are clear
- [x] Specification is complete with request/response formats
- [x] Backward compatibility addressed
- [x] Security implications considered
- [x] Reference implementations provided

SamMorrowDrums committed:
Add SEP: Dynamic Tool Annotations

SamMorrowDrums committed:
Formalize SEP: add PR number and rename file to 1862-dynamic-tool-annotations.md

findleyr commented:
By comparison, LSP has a notion of 'resolve', which is when the client asks the server to 'finish filling out' a particular stub. For example [`codeAction/resolve`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#codeAction_resolve), though this exists for many constructs.

That's slightly different, because the client isn't passing in any specific arguments in the resolve request: it's simply exchanging the stub for a fully filled-out version, but I think there's something to be learned from the design: we probably want to avoid a future where there are a bunch of separate preflight checks for a single tool call--it would be better to consolidate them into a single exchange. I'm not sure what that means, concretely: do we return the entire tool from the preflight check, rather than just its annotations?

SamMorrowDrums committed:
refactor(sep): consolidate examples, fix date, improve tone consistency

- Update creation date to 2025-11-21
- Consolidate duplicate file/API examples into single table-based example
- Remove redundant 'Why Return Full Tool?' section (already in Rationale)
- Clean up TypeScript comments
- Apply prettier formatting

SamMorrowDrums committed:
rename: SEP file to match new title (tool-resolution)

SamMorrowDrums committed:
Add future extensibility examples: auth delegation, execution characteristics, and related SEPs

- Add authorization delegation context example (based on SEP-214 token exchange)
- Add execution characteristics example for long-running task indication
- Update motivation to mention execution characteristics and auth boundary crossings
- Add SEP-1385 (Tool Execution Requirements) to Related Work
- Add SEP-214 (On-Behalf-Of Token Exchange) to Related Work

SamMorrowDrums committed:
Polish: reduce redundancy, improve clarity, add SEP links

SamMorrowDrums committed:
Delete seps/1862-dynamic-tool-annotations.md

SamMorrowDrums committed:
Integrate SEP-1488 (securitySchemes) with Tool Resolution

Adds argument-specific scope requirements as a key use case:
- Motivation: Over-scoping problem (e.g., GitHub repo scope for public repos)
- Future Extensibility: Dynamic securitySchemes example with noauth/oauth2
- Related Work: Cross-reference to SEP-1488 and SEP-711 (Trust Annotations)

The key insight is that static securitySchemes declarations require
maximum permissions upfront, but tool resolution enables servers to
return argument-specific requirements (e.g., noauth for public repos,
repo scope only for private repos).

This reduces unnecessary OAuth prompts and enables principle of least
privilege in MCP tool invocations.

SamMorrowDrums committed:
Format with prettier

SamMorrowDrums committed:
Fix prettier formatting

localden commented:
@SamMorrowDrums this SEP does not have an [assigned sponsor](https://modelcontextprotocol.io/community/sep-guidelines#sep-workflow). Are you looking to continue working on it and find sponsorship?

SamMorrowDrums commented:
Very much so @localden, it underpins several things I think are important when needed, I have got a few folks interested (including some major companies in the space) but I'm not sure how best to identify a sponsor.

If you have any advice I'd love to hear it. I'll bring it up on discord, perhaps I just need to work out which group would care most on discord or propose an IG, if there isn't one.

I'm going to have some opportunity in the Tool Scopes WG to see if it's a good fit there shortly, given that there isn't a 1:1 mapping with scopes and tools and it's not clear until runtime what the concrete requirement would be, it could be adopted for this.

I believe it would make the trust and privacy SEP proposal better too, reduce over prompting and give MCP clients a better UX.

I discussed its possible applicability to FGP with @nbarbettini last week also.

Need a little more time, and might need to tighten up the elevator pitch to encourage more interest.

SamMorrowDrums committed:
Update sponsor in 1862-tool-resolution.md

SamMorrowDrums committed:
Improve SEP-1862 elevator pitch and update to latest template format

- Update header to match SEP-1850 PR-based workflow format
- Emphasize key design principles: opt-in, backward compatible, resilient fallback, lightweight
- Add confirmation fatigue framing in consequences section
- Strengthen performance requirements (milliseconds, like HTTP OPTIONS/HEAD)
- Add extensibility messaging for auth/security/trust/privacy specifications
- Cross-reference SEP-1913 (Trust and Sensitivity Annotations)
- Add curated App Store use case showing layered approval model
- Regenerate SEP documentation for docs site

SamMorrowDrums committed:
Update sponsor handle from @nicknotfun to @nickoai

SamMorrowDrums commented:
> "Clients MAY skip resolution for tools where static metadata is acceptable" -- I'm not sure what this means

What I intended here is that if a client wasn't going to prompt a user anyway (because for example they enabled skipping confirmations for a given tool, all write tools or they're running yolo mode), the result of the resolution would be immaterial to the client and therefore could be reasonably skipped. 

The static annotations are meant to represent the worst case and so clients should be free to determine if there would be any value in the call.

The caveat to this is if we were to one day extend tools/resolve to support additional actionable data or behaviours such as issuing a scope challenge, it would then be better if clients always called tools/resolve when offered.

Do you have a suggestion for rephrasing @connor4312 or do you think we should just remove this part?

gyrgy commented:
My MCP server is per-action by design: separate tools for read, list, remove rather than one tool with an action argument. 

Part of the reason was the static-annotation worst-case problem this SEP describes. A multi-action tool can't honestly declare anything but `destructiveHint: true`, which makes it look more dangerous than most of its calls actually are. Resolution would change that calculus and unlock the multi-action design pattern.

On the open questions, from running 48 tools at this scale:

On Connor's question about server-capability vs per-tool annotation: per-tool. With 48 tools, half don't benefit from resolution at all because their behaviour doesn't depend on arguments. A resolve-able flag at the tool level keeps the implementation cost where the actual value is.

On "Clients MAY skip resolution for tools where static metadata is acceptable": this gets concrete if you collapse the existing booleans into a tier. 

- READ (`readOnlyHint=true`)
- MUTATE (`readOnlyHint=false`, `destructiveHint=false`)
- DESTRUCTIVE (`readOnlyHint=false`, `destructiveHint=true`)

Useful client heuristic: skip resolve for READ, always call for DESTRUCTIVE, optional for MUTATE. That gives "where static is acceptable" something to point at.

One thing worth surfacing in the rationale: `tools/resolve` improves the UX but doesn't change the trust model. A server that lies in `tools/list` can also lie in `tools/resolve`. The value is in giving honest servers a way to be more accurate, not in making annotations enforceable. Worth saying explicitly so adopters don't read the SEP as a security feature.
